clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -511,17 +511,18 @@ private static int parseAcks(String acksString) {
      * @return The amount of time we waited in ms
      */
     private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
-        // add topic to metadata topic list if it is not there already.
-        if (!this.metadata.containsTopic(topic))
-            this.metadata.add(topic);
-
+        // add topic to metadata topic list if it is not there already and reset expiry
+        this.metadata.add(topic);
         if (metadata.fetch().partitionsForTopic(topic) != null)
             return 0;
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            // Add the topic to metadata before requesting update to ensure that the topic is included
+            // in the metadata topic list since unknown topics are removed from the metadata
+            this.metadata.add(topic);
```

Comment:
@hachikuji This PR addresses two issues that were affecting the Confluent REST proxy. It attempts to reduce the impact of these issues without changing the current behavior of the producer.
1. A message is published to a topic using the REST proxy. For the rest of the lifetime of the proxy, metadata is refreshed for the proxy. This is an unnecessary overhead and a memory leak. The PR expires the topic to fix this.
2. A message is published to a topic using the REST proxy and after the publish the topic is deleted. For the rest of the lifetime of the proxy, the producer attempts to get metadata for the topic filling up the logs with errors since the topic has been deleted. With this PR, since there are no outstanding messages, the error is printed once and the unknown topic is deleted from the metadata.

Two other scenarios:
3. Create a topic and publish to it immediately. Producer attempts to get metadata until it is available (for the metadata refresh timeout). Errors are logged while the topic is being created, but the message will be published when the topic is available. This behavior is unchanged with this PR.
4. Publish a message to a deleted/non-existent topic. This is similar to the scenario above. Errors are logged as the producer attempts to get metadata for the topic. In this case, with this PR, errors are logged until the message expires. Without the PR, errors were logged indefinitely. 

I think your suggestion is to improve 4) to reduce the errors. Or tolerate more errors in 2) to keep the implementation simpler. I couldn't think of a way to improve 4) without impacting 3). And I am keen on fixing 2) since that is a lot of errors generated in the log for a scenario where there is actually no error at all (unlike 4) which is a user error since messages are being sent to a non-existent topic).